facebook/metro
# Add support for .mjs and .cjs file extensions

**Summary**

Node.js have added support for the `.mjs` and `.cjs` extensions which many packages have started using. This pull request adds these two file extensions so that packages using them can still be loaded with Metro.

This was specifically prompted by `@apollo/client`, a popular package that can be used from React Native, changing the file extension to `.cjs` which broke compatibility with Metro.

> - [Relevant if you use Apollo Client with React Native] Since Apollo Client v3.5.0, CommonJS bundles provided by `@apollo/client` use a `.cjs` file extension rather than `.cjs.js`, so Node.js won't interpret them as ECMAScript modules. While this change should be an implementation detail, it may cause problems for the [Metro bundler](https://facebook.github.io/metro/) used by React Native, whose [`resolver.sourceExts`](https://facebook.github.io/metro/docs/configuration#sourceexts) configuration does not include the `cjs` extension by default.
> 
>   As a workaround until [this issue](https://github.com/facebook/metro/issues/535) is resolved, you can configure Metro to understand the `.cjs` file extension by creating a `metro.config.js` file in the root of your React Native project:
>   ```js
>   const { getDefaultConfig } = require("metro-config");
>   const { resolver: defaultResolver } = getDefaultConfig.getDefaultValues();
>   exports.resolver = {
>     ...defaultResolver,
>     sourceExts: [
>       ...defaultResolver.sourceExts,
>       "cjs",
>     ],
>   };
>   ``` 

Fixes #535
Fixes #59
Fixes #60

**Test plan**

In a temporary directory:

1. `npx react-native@latest init TestSourceExts` 
2. `cd TestSourceExts`
3. `npx react-native run-ios`
    1. Check that project works
    2. Close everything opened
4. `yarn add @apollo/client graphql`
5. Add `import { ApolloProvider } from '@apollo/client'` to top of `app.js`
6. Add the following snippet to the `<App />`
    ```js
    <Section title="Apollo Import Test">
      Should be function: {typeof ApolloProvider}
    </Section>
    ```
7. `npx react-native run-ios`
    1. Observe the error
    2. Close everything opened
8. Manually patch `node_modules/metro-config/src/defaults/defaults.js`
    ```diff
    - exports.sourceExts = ["js", "json", "ts", "tsx"];
    + exports.sourceExts = ["js", "json", "ts", "tsx", "mjs", "cjs"];
    ```
9. `npx react-native run-ios`
    1. Observe everything gloriously working
    2. Celebrate 🎉
